deployment/elasticsearch/roundtripper.go

```diff
@@ -0,0 +1,74 @@
+package elasticsearch
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	v4 "github.com/aws/aws-sdk-go-v2/aws/signer/v4"
+)
+
+// elasticsearchRoundTripper implements RoundTrip to use it as a Transport in
+// an http client. It signs the requests using AWS Signature Version 4 before
+// letting the request go through its inner transport's RoundTrip.
+type elasticsearchRoundTripper struct {
+	signer    *v4.Signer
+	creds     aws.Credentials
+	region    string
+	transport http.RoundTripper
+}
+
+type DialContextF func(context.Context, string, string) (net.Conn, error)
+
+func newElasticsearchRoundTripper(dialCtxt DialContextF, creds aws.Credentials, awsRegion string) (*elasticsearchRoundTripper, error) {
+	signer := v4.NewSigner()
+
+	// Use the default transport, except for DialContext, for which we use the
+	// provided function, effectively tunneling all requests through that connection
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.DialContext = dialCtxt
+
+	return &elasticsearchRoundTripper{
+		signer:    signer,
+		creds:     creds,
+		region:    awsRegion,
+		transport: transport,
+	}, nil
+}
+
+// RoundTrip implements the RoundTripper interface, signing the request with AWS
+// Signature Version 4 before passing it to the underlying transport's RoundTrip
+func (s elasticsearchRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	if err := s.signRequest(req); err != nil {
+		return nil, err
+	}
+	return s.transport.RoundTrip(req)
+}
+
+// signRequest sign the provided request using AWS Signature Version 4
+func (s elasticsearchRoundTripper) signRequest(req *http.Request) error {
+	body := []byte{}
+	if req.Body != nil {
+		var err error
+		body, err = io.ReadAll(req.Body)
```

Comment:
I assume requests will be fairly minimal in terms of body size so this should be good enough.

Reply:
Yeah, at first I was going to avoid the ReadAll, but I don't think we'll get to a point where that's a problem, to be honest.